include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -268,10 +268,14 @@ class Matrix
           Matrix<ValueType, LocalIndexType, GlobalIndexType>>,
       public ConvertibleTo<
           Matrix<next_precision<ValueType>, LocalIndexType, GlobalIndexType>>,
+#if GINKGO_ENABLE_HALF
+      public ConvertibleTo<Matrix<next_precision<next_precision<ValueType>>,
```

Comment:
I think they need to use two from the logic of friend class.
`Matrix<value_type>: public ConvertibleTo<Matrix<next_precision>>`
with `friend Matrix<next_precision>` does not work. The original without half works because `next = prev`.
`Matrix<value_type>::convert_to(next_precision)` requires `Matrix<next_precision>` declare the `Matrix<value_type>` as a friend.
That is, matrix should use `friend Matrix<prev_precision>` to make `convert_to(next_precision)` work.
Thus, `convert_to next<next<>>` and `friend prev<prev<>>` are paired. `convert_to next<>` and  `friend prev<>` are paired.

Reply:
But it is defined as `previous_precision = next_precision<next_precision<T>`, so it should be equivalent?

Reply:
no, previous_precision is mapped to next_precision here.
if you delete friend next<next>, you need to delete function prev<prev> not next<next>
It is the confusing in the original code because the direction between friend and function is not consistent.